src/main/kotlin/org/rust/lang/core/macros/RangeMap.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.macros
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.util.SmartList
+import org.rust.lang.core.psi.RsMacroArgument
+import org.rust.lang.core.psi.RsMacroCall
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.expansion
+import org.rust.lang.core.psi.ext.startOffset
+import org.rust.stdext.optimizeList
+import org.rust.stdext.readVarInt
+import org.rust.stdext.writeVarInt
+import java.io.DataInputStream
+import java.io.DataOutputStream
+
+// TODO optimize it using sorted array and binary search
+class RangeMap private constructor(private val ranges: List<Pair<TextRange, TextRange>>) {
+
+    fun mapOffsetFromExpansionToCallBody(offset: Int): Int? {
+        return ranges.singleOrNull { (_, s) ->
+            offset >= s.startOffset && offset < s.endOffset
+        }?.let { (d, s) ->
+            d.startOffset + (offset - s.startOffset)
+        }
+    }
+
+    fun mapOffsetFromCallBodyToExpansion(offset: Int): List<Int> {
+        return ranges.filter { (s, _) ->
+            offset >= s.startOffset && offset < s.endOffset
+        }.map { (s, d) ->
```

Comment:
Let's use more meaningful names for ranges instead of `s` and `d` and use the same name in `mapOffsetFromExpansionToCallBody` and `mapOffsetFromCallBodyToExpansion`